File: InterceptingFilter/InterceptingFilter.go
package main

import "fmt"

type Filter interface {
	execute(request string)
}

type AuthFilter struct {

}

func(a *AuthFilter)execute(request string){
	fmt.Println("Authing Reqeust:",request)
}

type DebugFilter struct {

}

func(a *DebugFilter)execute(request string){
	fmt.Println("log Reqeust:",request)
}

type Target struct {

}

func(a *Target)execute(request string){
	fmt.Println("executing Reqeust:",request)
}

type ChainFilter struct {
	filters []Filter
	target *Target
}

func(c *ChainFilter)setTarget(target *Target){
	c.target=target
}

func(c *ChainFilter)addFilter(filter Filter){
	c.filters=append(c.filters,filter)
}

func(c *ChainFilter)execute(request string){
	for _,filter:=range c.filters{
		filter.execute(request)
	}
	c.target.execute(request)
}


type FilterManager struct {
	filterChain *ChainFilter
}

func(f *FilterManager)init(target *Target){
	f.filterChain=new(ChainFilter)
	f.filterChain.setTarget(target)
}

func(f *FilterManager)setFilter(filter Filter){
	f.filterChain.addFilter(filter)
}

func(f *FilterManager)filterRequest(request string){
	f.filterChain.execute(request)
}


type Client struct {
	filterManager *FilterManager
}

func(c *Client)setFilterManager(filterManager *FilterManager){
	c.filterManager=filterManager
}

func(c *Client)sendRequest(request string){
	c.filterManager.filterRequest(request)
}


func main(){
	filterManager:=new(FilterManager)
	filterManager.init(new(Target))
	filterManager.setFilter(new(AuthFilter))
	filterManager.setFilter(new(DebugFilter))

	client:=new(Client)
	client.setFilterManager(filterManager)
	client.sendRequest("HOME")
}

File: InterceptingFilter/InterceptingFilter.js
/**
 * Created by jamesyan on 2017/3/26.
 */

var AuthFilter={execute:function(request){console.log("Authing Request:"+request)}}
var DebugFilter={execute:function(request){console.log("log Request:"+request)}}
var Target={execute:function(request){console.log("executing Request:"+request)}}

var FilterChain=function(target){
    this.filterChains=[];
    this.target=target;
}
FilterChain.prototype.addFilter=function(filter){
    this.filterChains.push(filter);
}

FilterChain.prototype.setTarget=function(target){
    this.target=target;
}

FilterChain.prototype.execute=function(request){
    for(var f in this.filterChains){
        var filter=this.filterChains[f]
        filter.execute(request)
    }
    this.target.execute(request)
}

var FilterManager=function(){
    this.filterChain=new FilterChain(Target);
}
FilterManager.prototype.setFilter=function(filter){
    this.filterChain.addFilter(filter)
}

FilterManager.prototype.filterRequest=function(request){
    this.filterChain.execute(request)
}

var Client=function(filterManager){
    this.filterManager=filterManager;
}

Client.prototype.sendRequest=function(request){
    this.filterManager.filterRequest(request);
}